keps/1282-requeue-strategy/README.md

```diff
@@ -0,0 +1,306 @@
+### Goals
+
+Allow administrators to configure requeuing behavior based on the reason the Workload is being requeued.
+
+### Non-Goals
+
+## Proposal
+
+Make queue placement after preemption configurable at the level of Kueue configuration.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Consider the following scenario:
+
+* A ClusterQueue has 2 ResourceFlavors.
+* Kueue has admitted a workload on ResourceFlavor #2.
+* There is a stock-out on the machine type needed to schedule this workload and the cluster autoscaler is unable to provision the necessary Nodes.
+* The workload gets evicted by Kueue because an administrator has configured the `waitForPodsReady` setting.
+* While the workload was pending capacity freed up on ResourceFlavor #1.
+
+In this case, the administrator would like the evicted workload to be requeued as soon as possible on the newly available capacity.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+
+## Design Details
+
+### API Changes
+
+Add additional fields to the Kueue ConfigMap to allow administrators to specify what timestamp to consider during queue sorting based on the reason for eviction.
+
+Possible settings:
+
+* `UseEvictionTimestamp` (Back of queue)
+* `UseCreationTimestamp` (Front of queue)
+
+```yaml
+kind: ConfigMap
+metadata:
+  name: kueue-manager-config
+  namespace: kueue-system
+data:
+  controller_manager_config.yaml: |
+    apiVersion: config.kueue.x-k8s.io/v1beta1
+    kind: Configuration
+    # ...
+    requeueStrategy:
+      priorityPreemption: UseEvictionTimestamp
+      podsReadyTimeout: UseCreationTimestamp
+```
+
+### Changes to Queue Sorting
+
+#### Existing Sorting
+
+Currently, workloads within a ClusterQueue are sorted based on 1. Priority and 2. Timestamp of Creation/Eviction.
+
+```go
+package queue
+
+// ...
+
+func queueOrdering(a, b interface{}) bool {
+	objA := a.(*workload.Info)
+	objB := b.(*workload.Info)
+	p1 := utilpriority.Priority(objA.Obj)
+	p2 := utilpriority.Priority(objB.Obj)
+
+	if p1 != p2 {
+		return p1 > p2
+	}
+
+	tA := workload.GetQueueOrderTimestamp(objA.Obj)
+	tB := workload.GetQueueOrderTimestamp(objB.Obj)
+	return !tB.Before(tA)
+}
+```
+
+```go
+package workload
+
+// ...
+
+func GetQueueOrderTimestamp(w *kueue.Workload) *metav1.Time {
+	if c := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); c != nil && c.Status == metav1.ConditionTrue && c.Reason == kueue.WorkloadEvictedByPodsReadyTimeout {
+		return &c.LastTransitionTime
+	}
+	return &w.CreationTimestamp
+}
+```
+
+#### Proposed Sorting
+
+Add types that contain information about queue ordering strategy to the `apis/config/<version>` package. For example:
```

Comment:
I feel this goes to deep into implementation details. Just stating that the ordering is configurable, and how should be enough, typically. Unless there is some important reason to include the code here.

Reply:
Good callout, I ended up implementing this in the ClusterQueue in the WIP PR #1318.

Reply:
Indeed, most pieces of code should go away. Generally KEPs just need to describe the API surface and give at most pseudocode for the implementation.